### ISSUE TITLE
Avoid re-creation/evaluation of VDOM subtrees

### ISSUE THREAD

**Issue by josephsavona** [cyclejs/core#204]
I was reading over Cycle's implementation and noticed an unexpected behavior while trying out some of the examples: it seems that components are re-created any time their parent changes, causing significant portions of the reactive computation graph to be discarded and recreated more often than seems necessary. This is easier to explain by example.

### Example: TodoMVC

1. Open the [Cycle todomvc example](https://github.com/cyclejs/todomvc-cycle/) example app.
2. Create two new todos.
3. Breakpoint the [`TodoItem`](https://github.com/cyclejs/todomvc-cycle/blob/master/src/components/Todos/view.js#L78) component definition. 
4. Make *any* change to *any* todo, and observe that *all* TodoItem components are discarded and re-rendered from scratch.

Conceptually, after step 2 the graph of reactive values looks something like:

```
todo a -----\
             todos ---> app
todo b -----/
``` 

Based on my understanding of the documentation, if only todo "a" changes `TodoItem` would not be re-evaulated for todo "b", and that portion of the graph to be reused. In practice, this isn't the case and both components are recreated and re-rendered (to virtual dom). I also expected that adding a new todo would not cause todo's "a" and "b" to be created or even re-rendered, but they are.

### Generalization

More generally, this problem would seem to occur in *any* Cycle app that follows the documented patterns for constructing applications. In the [component documentation](http://cycle.js.org/components.html), for example, the BMI calculator's view function is as follows:

```javascript
const vtree$ = Observable.combineLatest(sources.props$, value$,
    (props, value) =>
      div('.labeled-slider', [
        // ... abbreviated
        input('.slider', {
          type: 'range', min: props.min, max: props.max, value
        })
      ])
  );
```

Because `input` immediately invokes the function and creates the component, there doesn't seem to be any way for an application to reuse a previous `input` at the same point in the graph (short of some memoization on the side).

### Questions

This seems like a potentially significant performance problem for applications. If `TodoItem` were instead a complex component - i.e. with a large number of (indirect) subcomponents - it would be expensive to re-construct frequently. Note that React solves this problem by using descriptors - the equivalent of the BMI `input` call in React would be to create a description of the input component and its arguments, which would allow React to reuse the existing component if it existed based on the DOM structure and `key`. Cycle DOM theoretically supports this same optimization - it's supported by the underlying `virtual-dom` diffing algorithm - but this logic exists only in the driver, which is abstracted from the view. It isn't immediately clear how a Cycle app could be changed to take full advantage of `virtual-dom`'s component reuse in order to avoid recreating the components as React would.

Some questions:
- Is this an intentional design decision? 
- What are the benefits of always recalculating the components?
- What is the suggested pattern for avoiding this type of reevaluation of components should it become a performance problem?

I don't mean to attack Cycle by posting this issue, and I should note that this same problem can occur in React applications as well. I once debugged an issue where a complex React component was constantly re-rendering, taking 100s of milliseconds each time, all because of a callback prop being reallocated by its parent on every render. Fixing this meant that the component's `shouldComponentUpdate` worked again and that rendering could be skipped unless it was strictly necessary. I bring this up to point out that performance problems can occur in any architecture, and I'm wondering what the equivalent of React's `shouldComponentUpdate` performance hook is in Cycle.

**Comment by staltz** [comment]
Good questions. Before getting into specifics, it's good to mention that Cycle has not had that much focus on performance. It is only recently (when API has stabilized) that we started doing performance improvements. One of the biggest issues, event handler re-attachment, is now solved with the use of event delegation pattern. The issues after that are started to be addressed.

That said, there are some misunderstandings on some particular issues.

**Regarding TodoMVC implementation**. It indeed renders each todo component every time any other todo component changes. But it does not have to be like that and this is not a decision by design. Unfortunately I don't maintain TodoMVC that often, so it's not the best piece of example app we have. Here is one example code where we make component instances independent by creating them when their corresponding data item in the model is created: 
https://github.com/cyclejs/cycle-examples/blob/master/advanced-list-nest/src/app.js#L28-L33

This technique solves the problem you mentioned.

> Because input immediately invokes the function and creates the component, there doesn't seem to be any way for an application to reuse a previous input at the same point in the graph (short of some memoization on the side).

`input` generates a virtual DOM element, in other words "VTree" (in React it would be ReactElement). This is a cheap object to create and this is simply the Virtual DOM approach. The whole point is to be able to cheaply re-create description objects without having to reuse the previous. `input` is not a component, it's a virtual DOM element.

On the last issue you mentioned, regarding `shouldComponentUpdate`: Cycle is not comparable to React. Cycle is comparable to React + Flux. It's the whole package. Not just that, it's that, but parameterized. Meaning, you can use Cycle as `virtual-dom` + MVI or Cycle as https://github.com/paldepind/snabbdom + MVI or Cycle as React Native + MVI, etc.

Which is to say: the equivalent to `shouldComponentUpdate` depends on what virtual DOM library you are using. Cycle can be used with React, which means you could use `shouldComponentUpdate`. Cycle DOM is the standard solution, and uses `virtual-dom`. The equivalent of `shouldComponentUpdate` in `virtual-dom` is [thunks](https://github.com/Matt-Esch/virtual-dom/blob/master/docs/thunk.md). We plan to migrate Cycle DOM to Snabbdom, and there, the option is [thunks](https://github.com/paldepind/snabbdom#thunks) too. If there happens to be some other virtual DOM library supported in Cycle, it will have yet another optimization technique. There has even been an experimental *Incremental DOM* driver for Cycle.js. So don't think of Cycle as an alternative to React. Cycle is a framework for structuring your app with streams, separating logic from pluggable effects.

All that said, I appreciate your thoroughly documented issue and we take this issue as a reminder to keep performance in mind more often.

**Comment by ronag** [comment]
Possibly related https://github.com/Reactive-Extensions/RxJS/issues/1032

**Comment by josephsavona** [comment]
Thanks for the detailed response! It's still a bit unclear how to resolve the original issue while also preserving the benefits of Cycle's approach (intents are really elegant!). It seems like there are two general approaches a developer could pursue:

#### Component-less Approach

> `input` generates a virtual DOM element, in other words "VTree" (in React it would be ReactElement). This is a cheap object to create and this is simply the Virtual DOM approach.

While `input` does create a virtual DOM element, there's no guarantee that creating it would be cheap. There's nothing stopping a developer from calling arbitrary functions that may produce complex virtual dom subtrees. For example, this is the approach you used in your [flux-challenge submission](https://github.com/staltz/flux-challenge/blob/master/submissions/staltz/src/view.js#L37-L47). With this approach, memoizing the results of previous calls to sub-view functions (`slots()`, `scrollButtons()`, etc) would be up to the user with thunks...

> The equivalent of shouldComponentUpdate in virtual-dom is thunks.

Cycle DOM treats thunks as [raw virtual dom nodes](https://github.com/cyclejs/cycle-dom/blob/354fd5ab404a5ff220b3737c10e5852add7fdbab/src/transposition.js#L22-L26), i.e. it does not expand them and compose any nested Observables they may contain. Since `virtual-dom` (and presumably Snabbdom) cannot directly render Observable values, using them to solve performance at the product level would seem to preclude the use of actual subcomponents and Cycle's intent system.

#### Component Approach

> Here is one example code where we make component instances independent by creating them when their corresponding data item in the model is created:

I see, this helps to clarify the architecture quite a bit. To confirm, to achieve subcomponent reuse parent components should avoid the obvious implementation:

```javascript
function TodoList(todos$) {
  return {
    DOM: todos$.map(todos => (
      // this reallocates all `TodoItem`s every time the input `todos$` changes
      div('#todos', null, todos.map(todo => (...)) 
    )),
  };
}
```

Instead, applications would have to memoize previously constructed children in the component "constructor" so as to reuse existing children:

```javascript
function TodoList(todos$) {
  const todoItemMap = new Map(); // 
  const memoizedTodos = todos$.map(todos => {
    // Output a list of `TodoItems`, but use a Map or something to memoize the `TodoItem`
    // components created between executions. For example, the Rx equivalent of: 
    todos.forEach(todo => {
      if (!todoItemMap.has(todo.id)) {
         todoItemMap.set(todo.id, TodoItem(todo));
       }
       todoItem = todoItemMap.get(todo.id);
    });
  });
  return {
    DOM: memoizedTodos$.map(memoizedTodos => (
      div('#todos', null, memoizedTodos) 
    )),
  };
}
```

In either case, the two obvious approaches seem to require the developer to manually implement sub-component re-use. That isn't necessarily bad, but it's an interesting side effect of the architecture of the system.

> All that said, I appreciate your thoroughly documented issue and we take this issue as a reminder to keep performance in mind more often.

Thanks again for the great response!

**Comment by ronag** [comment]
Just to shime in a little bit. If you look at the `resolve` operator (which although a bit badly implemented would improve your example:

```javascript
function TodoList(todos$) {
  return {
    DOM: resolve(todos$, todo => (...)).map(todos => (
      div('#todos', null, todos) 
    )),
  };
}
```

Where the todos would be reused automatically.

**Comment by milankinen** [comment]
Another strategy is to define your data as a stream of streams and use `flatMapLatest` + `combineLatest` combo -- every time when your (todo) item changes, `combineLatest` uses the latest ("cached") values from other streams, thus re-rendering only the VTree of the modified item. 

If the list changes (e.g. adding/removing items), then the VTree of each item must be re-rendered. Of course also this can be tackled by using André's strategy. Another benefit in this approach it the better isolation of sub-applications - one can trivially memoize the function which makes possible to do expensive initializations (e.g. ajax calls) inside the memoized function. Of course this approach has also its drawbacks: your data model must be a multicast hot observable so that it doesn't loose its "state" during the full re-renders. 

Demo here (see console logs to understand the behaviour): https://jsfiddle.net/milankinen/gbsg55tc

That said, modifications must still propagate to the DOM driver which does the vdom diffing - and this might be a possible bottleneck when dealing with extremely deep vtrees (subscription overhead) or large item sets (diffing overhead), even though the leaf vtrees can be memoized/cached.

**Comment by ronag** [comment]
@milankinen: Thank you! That's basically what I've been trying to do with `resolve` for a long time now. Your solution is much more elegant!

**Comment by niieani** [comment]
Parts of this conversation would make a good starting point for a "best practices" article for the docs. :+1:

**Comment by RoelVanderPaal** [comment]
how can I prevent the Child counter from being reset to 0 in this example?: https://jsbin.com/yezudot/edit?html,js,output

**Comment by Funx** [comment]
Use ```combineLatest``` instead of ```map``` in your main function: 

```javascript
function main(sources) {
  const counter$ = Rx.Observable.interval(1000);
  const child$ = Child().DOM;

  return {
    DOM: Rx.Observable.combineLatest(
      counter$,
      child$,
      (counter, child) => div([counter, child])
    )
  };
}
```

https://jsbin.com/nerudediju/edit?html,js,output

**Comment by josephsavona** [comment]
@staltz I'm curious why you decided to close this issue. I believe I've raised some legitimate concerns about the performance of Cycle in common use cases, pointing out numerous instances of unnecessary re-computation in the example applications. I believe the pattern of performance problems is quite clear, but so far all of the solutions proposed are highly specific. 

Do you plan to address any of these concerns?

**Comment by ronag** [comment]
@josephsavona: I second that.

**Comment by ronag** [comment]
https://github.com/cyclejs/core/issues/244 another example that also has this issue

**Comment by TylorS** [comment]
Please continue having this discussion. The new policy for discussions are to close them as they are not actual issues.

See https://gitter.im/cyclejs/core?at=56cf2c441ab7b28c4f360d18

**Comment by milankinen** [comment]
More generic approach could be to properly isolate the DOM driver (instead of the current silly CSS class hack). Since the current driver uses `virtual-dom`, it's **[fairly straightforward](https://gist.github.com/milankinen/573fcf4c28b6c4225386)** to isolate the subtree rendering by using `Widget`. In practice, this'll stop the vdom update propagation outside of the isolated component after "initial render". 

See the difference in practice (from developer console):
* Official: https://jsfiddle.net/milankinen/mk42tqc9
* Modified: https://jsfiddle.net/milankinen/Lsbcpna7

Because in this approach the root (DOM) node of the returned subtree handles the event propagation, it's possible to get rid of the current CSS class based isolation, possibly (at least partially) resolving #226 and #226. And because event listeners of the isolated component are bound to *real* child DOM node, also #229 might be solved.

However, this solution has also its drawbacks: 

In order to achieve "complete" isolation, every instance of the component must be isolated separately (at least with the current `@cycle/isolate` implementation) or otherwise every component gets events from all other components too (multi-instance support not implemented in my example). 

The second drawback is that dom sink just "stops" emitting vtree changes, thus making post-dom transforms (like animations) impossible or at least very challenging.

**Comment by ronag** [comment]
@josephsavona: Here is my solution for this problem.

Using: https://gist.github.com/ronag/3db1f020416b895f7519

You can do:

```javascript
import { resolve, resolveCombineLatest } from './rx-resolve'

function TodoList(todos$) {
  const todos$ = resolve.call(todos$, todo => (...))
  return {
    DOM: resolveCombineLatest.call(todos$, todo => todo.DOM, todos => 
      div('#todos', null, todos.map(todo => (...)) 
    ),
  };
}
```

This way nothing will be unnecessarily re-allocated.

**Comment by josephsavona** [comment]
@ronag Thanks for the link, the resolve* helpers seem really useful. It would be great to see the Cycle examples updated to encourage this pattern.

**Comment by staltz** [comment]
Just checking in to say it's worth labeling and re-opening this as a real issue, not just a discussion. Also renamed the title a bit. We indeed could put this type of optimization in the library as Ronag and Joseph wrote above.

I don't have much time to give every issue full attention (I don't work on Cycle.js full time), but I'm actively search for time to work on issues. Of course it'd be nice to have more contributors from the community.

**Comment by staltz** [comment]
I'd also like to comment that while unnecessary computation can be _avoided_, it's not a problem for performance until it actually expresses itself as a perceivable problem. We're working on getting the _overall_ (not just DOM concerns) performance up a lot, by using faster virtual DOM libraries (snabbdom and hopefully inferno) and faster stream libraries (RxJS v5 is ~5x faster than RxJS v4, most.js may be event ~300x faster than RxJS v4). After we do that, if unnecessary calculations in virtual DOM subtrees expresses itself as a perceivable and common problem, we'll start addressing this issue.

PS: when I say "we", it's me and @TylorS, and hopefully @Frikki too when he's back.

**Comment by ronag** [comment]
@staltz: For me this is not primarily a DOM issue. It's a big issue when there are components that need to perform side effects in general, e.g. a component that needs to fetch it's data using a HTTP request. You don't want to have to re-do those...

**Comment by ronag** [comment]
But yes, this is not really a cycle issue. It's more of a rx missing feature. I've had the same issue in other contexts as well.

**Comment by ronag** [comment]
@staltz: Just to clarify, given the optimisation you mention, do you think my suggestion is a bad idea?

**Comment by ronag** [comment]
I've opened an issue on RXJS (https://github.com/Reactive-Extensions/RxJS/issues/1145). Hopefully this can be a generic thing...

**Comment by ivan-kleshnin** [comment]
I think this problem is much more general than being just about "component rendering".
The question is about granularity of reactive state.

All CycleJS examples I ever saw imply that state is a single stream.
Which does not scale for real-world cases.

```js
// "snapshot" streams
let state = Observable.of({
  users: [],
  items: [],
  // blah-blah-blah
})
```

This has two fundamental problems.

First: **to update state we need to apply complex selectors**.

Second: **we can't depend on a state fragment**.

```js
state.subscribe((state) => console.log(state.users)) // do something with state.users
// is updated every time state updates... :(
```

What we really want is

```js
// "snapshot" streams
let state = {
  users: Observable.of([]),
  items: Observable.of([]),
  // blah-blah-blah
}
```

Then code parts can depend on `users` being ignorant of `items` and so on.

```js
state.users.subscribe((users) => console.log(users)) // do something with state.users
// is updated only when users update... :)
```

To be able to do this **@cycle/core** looping mechanics shoud be updated.

If plain object is passed instead of `Observable` the tree of streams (namespaced streams in other words) should be assumed and proxy should be created for each tree **leave** instead.

I make a partial solution [here](https://github.com/ivan-kleshnin/cyclejs-examples/blob/master/1.5-form/src/rx.utils.js#L40-L54) until I've realized this should rather belong to **@cycle/core**. But the algo direction with flatten/unflatten is probably right.

**Comment by niieani** [comment]
I agree with @ivan-kleshnin on that we should have the possibility to pass a tree of streams. I've run in to the same problem myself.

**Comment by polytypic** [comment]
@ivan-kleshnin I don't really feel welcome to comment here, but in Calm^2 we solve the problem of transmitting state to components using a powerful combination of modifiable reactive properties, called Atoms, and bidirectional computations, called lenses.  Indeed it makes Calm^2 components nestable (or "fractal") and is both a really concise and a very expressive way to do it.  Note that Calm^2 is quite different from Cycle.js, but is also an approach that is fundamentally based on observables (the most important ingredient). It simply chooses to structure things quite differently.  Calm^2 libraries currently exists for Bacon and Kefir, but it can be ported to pretty much any event stream combinator library, including Rx.

tl;dr; both of your problems 1 and 2 are *solved* problems in Calm^2.

**Comment by ivan-kleshnin** [comment]
@polytypic do you mean [calmm-js](https://github.com/calmm-js)? A lot of just-updated repos are there and I'm not really sure which one do you refer to. As I understood the whole project kit is not documented yet. I can't say for the commumity but I'd personally like to get some cues from your experience.

**Comment by polytypic** [comment]
@josephsavona Again, I know my comments aren't welcome here, but in Calm^2 the goal of avoiding unnecessary computation has been present right from the start.  The idea has been to make VDOM updates effectively *incremental* similar as to in SAC or Self-Adjusting Computation.  (BTW, SAC is a concept that I first run into in 2005 at ICFP / ML Workshop in Tallinn.)  Thanks to this, components in Calm^2 can be straightforwardly written in a way that VDOM updates are nearly minimal.  My Calm^2 and Kefir based implementation of Todo"MVC" seems to be both one of the fastest and one of the most concise implementations: https://github.com/calmm-js/kral-todomvc (click on the link to see a live version).  Here is a snippet of code you can enter into the console to populate the implementation with 2000 items (you must then refresh the page):

```js
var store = { list: [] };
for (var i = 1; i <= 2000; i++) {
  store.list.push({ title: 'Todo' + i, completed: false });
}
localStorage.setItem('todos-react.kefir', JSON.stringify(store.list));
```

@ivan-kleshnin Yes.  I'm currently in the process of documenting it further, but it will take time.

**Comment by staltz** [comment]
@polytypic to you remind you about our [code of conduct](https://github.com/cyclejs/core/blob/master/CODE-OF-CONDUCT.md):

> Likewise any spamming, trolling, flaming, baiting or other attention-stealing behavior is not welcome.

**Comment by ivan-kleshnin** [comment]
@Cmdv you proposed something very similar at https://github.com/cyclejs/core/issues/191
What do you think now? Our perspectives were a bit different but implementations are converging IMO.

**Comment by laszlokorte** [comment]
@ivan-kleshnin I think cycle-core does not need to be adjust for this to work. You just need a driver with lens support. Like this one: https://github.com/milankinen/stanga/blob/master/src/model.js
Then you can have a single observable that contains the global state but can be deconstructed into observables containing only a slice of the state

**Comment by niieani** [comment]
@laszlokorte Could you explain how do lenses work and what they are? A simple walkthrough of the model code would be enough. I've tried googling around, but couldn't find any information on that. Granularity of the state could greatly help me with my endeavor to write a cycle.js driver for the Aurelia framework.

**Comment by laszlokorte** [comment]
@niieani For a more in depth explanation you can take a look at this:
https://www.youtube.com/watch?v=efv0SQNde5Q

A lens is basically a path to a nested property of an object.
Say you have an object:
`{root: { children: [ {attrA: "bar", attrB: "foo"} ] }}`
Now you could read the attrA value via: `const myVal = yourObject.root.children[0].attrA`
But now if the object get's changed your myVal will not update.
If instead you just read the child `const child = yourObject.root.children[0]` you have a reference to the child object so if the attrA get's changed your object will have the new attrA value as well. But if somebody replaces the whole child you are still out of luck.

Now a lens represents that path `yourObject.root.children[0].attrA` without actually reading the value so you can read the value whenever you need it an get the most recent one.
eg:
`const myLense = lense("yourObject.root.children[0].attrA", yourObject)` and then whenever you call `myLense.get()` you will get the correct value no matter how `yourObject` has changed.
Same works for writing: you can call `myLense.set("barbar")` and the object will be updated correctly.

Now the real advantage is that you can pass somebody a lens without telling where it points. Eg I can pass you a `myCoolLense` object an you can read it and write to it without know what will actually happen. It could be referring to just one value at the root of my object or it could be referring to a value very deep nested inside my object hierarchy. All I tell you is that it's a lens containing a string. And you can construct lenses out of lenses. So If I construct a lens pointing to a user object inside my storage you can take that lense, knowing it "contains" a user and build a new lens that only point's to the user's name and pass that new lens on to somebody else - just telling her that it's the user's name. She can now go and `set` that's lens' value. That change will propagate back to you user object and to my storage.

**Comment by Cmdv** [comment]
@ivan-kleshnin I think that with pipping your state through out a Cycle app you actually end up getting the best of both worlds.

1. you can still if you wish encapsulate local state and you have the flexibility of the top down vs bottom up

2. I pipe my state back up the tree and through `run` using a  `.just()` observable as a driver in it's simplest form (bit more otherwise) so in fact I see the act of pipping everything back up in that location as being a single communal point for my state.

3. With point 2 in mind you still have the flexibility of knowing the whole app state with each iteration but you don't have to pass the whole object around. You can therefor decide how you would like your state to behave and it doesn't have to be just one piece.

4. You get a very good visual representation how the data-flow of your state in the code it self. you simply follow the sources & sinks of each component and you can map everything.

You do get gotcha's that you need to make sure state Observables are hot, and in my case needed to ability the state when changing route so it could be called upon again. This is the mini Driver @TylorS help build:

```
function makeStateDriver(stateStore = {}) {
  return function stateDriver(sink$) {
    const source$ = sink$
      .scan((prev, curr) => {
        if (!curr || typeof curr !== 'object') {
          throw new TypeError('state must be an object')
        }
        return Object.assign(prev, curr)
      }, stateStore)
      .do(x => console.log('makeStateDrive sinks$: ', x))
      .replay(null, 1)
    source$.connect()
    return source$
  }
}

export {makeStateDriver}
```

So even though the Redux is a really attractive solution and think it has it's limitation and you can easily replicate it if you chose to. For me Cycle really needs to be non subjective on that design choice, let the developer decide if that need it or not :)

**Comment by milankinen** [comment]
Nice explanation @laszlokorte!

Reactive lenses have two parts: get and set. Get is easier to understand. `lens(obs$, "foo")` is practically equal to `obs$.map(x => x.foo)`. That's nothing new. But the power of lenses becomes from `set`. Lens is kind of "zoom" - if you modified something in the "zoomed" area, also the original area gets the changes.

In Cycle, modifications to state are usually done by using `mod$` streams like:
```javascript
function Child({DOM}) {
  const modProxy$ = new Subject()
  const state$ = modProxy$.startWith([]).scan((s, mod) => mod(s)).shareReplay(1)
 
  const addItemMod$ = DOM.select(".add").events("click")
    .map(() => items => [...items, newItem(...)])

  modProxy$.subscribe(addItemMod$)
  return { DOM: state$.map(renderItems) }
}
``` 
The lens abstraction tries to remove the child component's "awareness" of the origin of the *real* state (`state$` above). What it sees, is just a lens (observable having `.lens` and `.modify`) and the modification must be done by using lens' `.modify` function.

```diff
function Child({DOM}) {
  const modProxy$ = new Subject()
+ const state$ = makeLens(modProxy$.startWith([]).scan((s, mod) => mod(s)).shareReplay(1))
- const state$ = modProxy$.startWith([]).scan((s, mod) => mod(s)).shareReplay(1)
 
  const addItemMod$ = DOM.select(".add").events("click")
    .map(() => items => [...items, newItem(...)])
+   .map(state$.modify)

  modProxy$.subscribe(addItemMod$)
  return { DOM: state$.map(renderItems) }
}
``` 
Complicated? Yes, but let's see what happens if we move the proxy to parent component:
```javascript
function Parent({DOM}) {
  const modProxy$ = new Subject()
  const state$ =  makeLens(modProxy$.startWith({items: []}).scan((s, mod) => mod(s)).shareReplay(1))
  
  const child = Child({DOM, state$: state$.lens("items"))

  modProxy$.subscribe(child.mod$)
  return {DOM: child.DOM}
}

function Child({DOM, state$}) {
  const addItemMod$ = DOM.select(".add").events("click")
    .map(() => items => [...items, newItem(...)])
    .map(state$.modify)

  return { DOM: state$.map(renderItems), mod$: addItemMod$ }
}
``` 
Okay. We can still move the proxies and state creation "upwards" - let's make it a driver (now we must rename `mod$` to `state$` because Cycle's `run` connects sources to sinks by driver's name). 

```javascript
function StateDriver(mod$, intialState = {items: []}) {
  return makeLens(modProxy$.startWith(initialState).scan((s, mod) => mod(s)).shareReplay(1))
}

run(Parent, {DOM: makeDOMDriver(...), state$: StateDriver})

function Parent({DOM, state$}) {
  const child = Child({DOM, state$: state$.lens("items"))
  return {DOM: child.DOM, state$: child.state$}
}

function Child({DOM, state$}) {
  const addItemMod$ = DOM.select(".add").events("click")
    .map(() => items => [...items, newItem(...)])
    .map(state$.modify)

  return { DOM: state$.map(renderItems), state$: addItemMod$ }
}
``` 
Components got a lot cleaner, huh? Let's add another child:
```diff
function StateDriver(mod$, intialState = {items: []}) {
+ return makeLens(modProxy$.startWith({a: initialState, b: initialState}).scan((s, mod) => mod(s)).shareReplay(1))
-  return makeLens(modProxy$.startWith(initialState).scan((s, mod) => mod(s)).shareReplay(1))
}

run(Parent, {DOM: makeDOMDriver(...), state$: StateDriver})

function Parent({DOM, state$}) {
+  // isolation is needed only for DOM driver, duh
+  const a = isolate(Child, "a")({DOM, state$: state$.lens("a").lens("items"))
+  const b = isolate(Child, "b")({DOM, state$: state$.lens("b").lens("items"))
+  const vdom$ = O.combineLatest(a.DOM, b.DOM).map(renderBoth)
-  const child = Child({DOM, state$: state$.lens("items"))
+  return {DOM: vdom$, state$: O.merge(a.state$, b.state$)}
-  return {DOM: child.DOM, state$: child.state$}
}

function Child({DOM, state$}) {
  const addItemMod$ = DOM.select(".add").events("click")
    .map(() => items => [...items, newItem(...)])
    .map(state$.modify)

  return { DOM: state$.map(renderItems), state$: addItemMod$ }
}
``` 
Also note that you can share the *same lensed state* between multiple components and they are kept sync all the time:
```diff
function Parent({DOM, state$}) {
+ // a and aCopy are identical and changes in other component are immediately visible in another one
  const a = isolate(Child, "a")({DOM, state$: state$.lens("a").lens("items"))
+ const aCopy = isolate(Child, "aCopy")({DOM, state$: state$.lens("a").lens("items"))
  const b = isolate(Child, "b")({DOM, state$: state$.lens("b").lens("items"))
+ const vdom$ = O.combineLatest(a.DOM, aCopy.DOM, b.DOM).map(renderAll)
+ return {DOM: vdom$, state$: O.merge(a.state$, b.state$, aCopy.state$)}
- const vdom$ = O.combineLatest(a.DOM, b.DOM).map(renderBoth)
- return {DOM: vdom$, state$: O.merge(a.state$, b.state$)}
}
```

And that's not all. Lenses enable you to do much much more fancy things like setting default values if the property does not exists, even selecting list components with [binary search](https://github.com/calmm-js/partial.lenses#food-for-thought-bst-as-a-lens).
```javascript
function Parent({DOM, state$}) {
  // isolation is needed only for DOM driver, duh
  const a$ = state$.lens(L("a", L.default({items: []})))
  const b$ = state$.lens(L("b", L.default({items: ["tsers!"]})))
  const a = isolate(Child, "a")({DOM, state$: a$.lens("items"))
  const b = isolate(Child, "b")({DOM, state$: b$.lens("items"))
  const vdom$ = O.combineLatest(a.DOM, b.DOM).map(renderBoth)
  return {DOM: vdom$, state$: O.merge(a.state$, b.state$)}
}
```
And lenses are also "fully fractal" (@staltz can confirm this).

**Comment by ivan-kleshnin** [comment]
@laszlokorte @Cmdv @milankinen so I see we're kinda agree that piping state through drivers is a good idea though the exact implementation is a question, right? I haven't got this impression of agreeness reading trhrough older threads. 

@laszlokorte but if we apply lensing mechanics to this task it can be used for others. Like to extract different DOM sinks from a sink record inside `@cycle/core`?! 

Then type of `main` can be tightened up from 

````
Observable a | {Observable *} -> Observable b | {Observable *}
```

to 

````
Observable a -> Observable b
```

This is the same stuff, isn't it?

**Comment by ivan-kleshnin** [comment]
> I think cycle-core does not need to be adjust for this to work. You just need a driver with lens support. Like this one: https://github.com/milankinen/stanga/blob/master/src/model.js
Then you can have a single observable that contains the global state but can be deconstructed into observables containing only a slice of the state

@laszlokorte if some driver implements lensing than state steam can't be inspected before passing it to sink?! This is something I'd like to have.

```js
function main({state: stateSource}) {
  // stateSource contains lensing methods (were added by driver)

  let stateSink = ...
  // stateSink does not contain lensing methods

  // ...so I can't create a stateSink derived Observable here...

  return {
    state: stateSink
  }
}

**Comment by laszlokorte** [comment]
> if some driver implements lensing than state steam can't be lensed before passing it to sink?! This is something I'd like to have.

Depends on how you create that `stateSink`. I think you would use `stateSource` for creating `stateSink`

**Comment by milankinen** [comment]
@ivan-kleshnin I didn't quite understand your question. Could you elaborate?

**Comment by ivan-kleshnin** [comment]
@laszlokorte soory I wasn't entirely precise. I mean that if we use lensing in some part of a `main` and basic `.filter()` fallback in other part – it's not good. I updated the code appropriately.

@milankinen that's because I'm thinking in classic CycleJS IMV paradigm while you apply your own architecture. I tried to follow your arguments for that one but failed, honestly. Your terms usage is alien to me (like using definition of transducer from electronics rather than Clojure / JS one).

**Comment by laszlokorte** [comment]
@ivan-kleshnin when creating a lens you always need to base it on some base lens in order to keep the whole chain. If you just create a lens `friends[0].name` without referencing the source `user` have were given how would you be able to read or write that lens? The link between your lens and the root of the state tree is missing. Not sure if this is addressing your question.

**Comment by ivan-kleshnin** [comment]
@laszlokorte yeah I know how lenses work for normal objects. I even [made a library](https://github.com/paqmind/data-lens) once ago. My question was about exposing this lensing machinery through a driver. If I failed to put it properly – so be it. I appreciate your explanations anyway. So I'm going to follow your advice and experiment with lensing as well.

**Comment by laszlokorte** [comment]
@ivan-kleshnin Yeah I assumed that you know how lenses work. I am just not sure what you suggest as solution for the issue you pointed out - if any?

**Comment by HighOnDrive** [comment]
Quick question, is it possible to change the data backing the defined lenes at runtime?

**Comment by ivan-kleshnin** [comment]
> I am just not sure what you suggest as solution for the issue you pointed out - if any?

@laszlokorte I just tried to say that it's (probably) not enough to return a lens-aware stream from a driver. 

@HighOnDrive no, at least not for classic van Laarhoven lenses. 
What's inside a closure is beyond your reach. You can only recreate a lens from scratch.

**Comment by HighOnDrive** [comment]
@ivan-kleshnin I'm appreciating your investigation (linked) and also the related links below. Was considering Cycle drivers and their usages again and it is really quite clear. Cycle is about dataflow and data flow happens on circuits, these are the main deal. Circuits are neutral and using them for handling side-effects or state management are just two usages of circuits. 

So it is natural that there would be different classes of drivers. The debate over doing a noop state-driver and if it fits into the usual use of circuits reminds me of the battle my neighbor is always dealing with. He heads the move to open bike (cycle) lanes on our city roads, more power to him!

Love these links:

[Reactive state implementations](https://github.com/ivan-kleshnin/reactive-states)
[calmm-js/partial.lenses](https://github.com/calmm-js/partial.lenses)
[The essential Cycling gear every Cyclist needs](https://github.com/milankinen/stanga#tutorial)

We better have bike lanes this summer!

**Comment by staltz** [comment]
> All CycleJS docs / examples I've seen implement a state as a single stream or imply it should be implemented so. Which does not scale for real-world cases.

Did this tangent start just because people want to avoid redundant computations?

If that's so, I don't think it's an actual problem, or at least not a big enough of a problem to prompt a significant architectural change.

Redundant computations have been harmless for logic. If anyone considers them harmful for performance, we are working on making Cycle's foundations (stream libraries and important drivers) faster, probably in the order of ~5x faster. Which means, if you had 5 redundant computations, by using the faster upcoming major version of Cycle, it's about the same as if you re-architected to avoid redundant computations.

When it comes to state object fragments as streams, one can also use distinctUntilChanged to avoid duplicates, e.g.: https://github.com/futurice/power-ui/blob/master/src/components/pages/PeoplePage/index.js#L28-L33

That said, a handy Lens library for Observables may help in this regard, but I would definitely not make a noop nor a state driver to achieve that, because that's not the point of drivers. Drivers are functions that take outgoing messages, produce side effects in the actual external world, and produce incoming messages. "Read effects" are messages **from** the external world, "write effects" are messages **to** the external world. That's why they are called drivers: they are a bridge between software and devices/effects/hardware. Internal app state has nothing to do with all this. Misusing drivers is not any better than dropping Cycle.js and using a different approach.

**Comment by milankinen** [comment]
It's not about performance (although that's bad too). It's about [unnecessary complexity](https://gist.github.com/milankinen/cb0e898ae52c61e8d5da#file-main-js) that could easily be avoided.

**Comment by HighOnDrive** [comment]
The sky is not just dominated by one kind of planet, they come in shapes and sizes, all contributing to the solar system on their own orbit. Same for circuits and what mission or payload they may have. Drivers is just a name for a usage of circuits, maybe a unique name for the new found use of circuits is all we need.

**Comment by ivan-kleshnin** [comment]
> It's not about performance (although that's bad too). It's about unnecessary complexity that could easily be avoided.

Yes. And about solving circular dependencies which can exist between computations.
Why should I use `Subjects` for that if we already have driver mechanics?
Only because we don't participated that and named them "drivers"? I don't think it's a big reason.

**Comment by Cmdv** [comment]
@staltz so a state driver is not recommended?
I tend to send my state/props through the driver of `run` because that is the only place I see that I know will feed it back out so it's present in my sources (`sources.props`) of my components.

Or would you recommend a high level state handler that sits say at the `main` top level?

**Comment by staltz** [comment]
noop drivers are simply in order to avoid using Subjects for circular deps.

> Why should I use Subjects for that if we already have driver mechanics?

Because JavaScript is a language where the order of declarations matter (not all languages are like this), and because using a proxy subject is a small price to pay (just two more lines of code).

The problem with using drivers for anything beyond external communication is not about lack of safety, it's about conflating concepts. This starts making it less clear what is a driver for. The more drivers are used for ad hoc solutions, the more confusion appears as to "when should I make a driver and when should I not?". Exceptions like the noop are shortcuts which make it harder for beginners to understand what do drivers do. "*well, you can safely use them as channels to transfer messages for whatever you want*" is something I'd be uncomfortable teaching to other programmers.

**Comment by staltz** [comment]
@Cmdv 
> so a state driver is not recommended?

Not recommended. Use proxy Subjects instead, like https://github.com/futurice/power-ui/blob/master/src/components/pages/PowerheadPage/index.js#L30. There is never a case where a "state driver" is the only possible solution.

**Comment by Cmdv** [comment]
@staltz ah ok, in the past I've done [this](https://github.com/Cmdv/cycle-webpack-boilerplate/blob/master/src/app.js#L22-L27)

which is a bad idea :( just felt natural to have all the data flows back up to the top then out.
But obviously an exposed proxy will do a similar job just not pass it through `run`.

Can we not have another, say hole to fit our pegs into through run that's specifically for state. Because would that not also be a good place to have all of the app info where time travel and things like that would be made easier or am I coming up with more anti patterns? :fearful: 

```
const drivers = {
  DOM: makeDOMDriver('#application'),
  History: makeHistoryDriver({hash: false, queries: true}),
};

const state$ = {
    Props1: () => Rx.Observable.just(0),
    Props2: () => Rx.Observable.just(0)
}

Cycle.run(main,drivers)(state$);
// or
Cycle.run(main,drivers,state$);
```

**Comment by ivan-kleshnin** [comment]
> The problem with using drivers for anything beyond external communication is not about lack of safety, it's about conflating concepts. This starts making it less clear what is a driver for. The more drivers are used for ad hoc solutions, the more confusion appears as to "when should I make a driver and when should I not?". Exceptions like the noop are shortcuts which make it harder for beginners to understand what do drivers do. "well, you can safely use them as channels to transfer messages for whatever you want" is something I'd be uncomfortable teaching to other programmers.

That makes sense and still I can't stop thinking that looping streams by `Cycle.run` is a good idea.
I wouldn't call `identity` usage for that matter an "exception case". It's an edge case. Like multiplying by one or adding zero... Store Drivers are entirely different thing. They clearly go against CycleJS assumptions but may be fine for Tsers or other alternative architectures people are trying out.

**Comment by ivan-kleshnin** [comment]
Why not just `Subject`!? 

If you make a 1-stream state it's not a big deal to loop it imperatively.
May not overweight the drawback of "conflating" concepts.

But in my example where I used granular state boilerplate is multiplied!

```
let update = {
  users: new Subject(),
  articles: new Subject(),
  ...
}

let state = {
  users: [],
  articles: [],
  ...
}

// --> wire up every state stream to every Subject channel <--
// i.e. N manual .subscribe()

return {
  ...
}
```

**Comment by HighOnDrive** [comment]
For sometime now I've see many attempts at grokking state management in Cycle, it was actually starting to look like React II with props and proxy Subjects are also along those lines. IMO, Cycle has ignored state management and focused mostly on the wonders of intent processing and projected state.

Circuits are what is unique and what Cycle should build upon in all possible ways. There are obviously classes for the functions that work on these circuits. Time has proven that that many developers still keep guessing if something should be a driver or not, so waving the name "driver" around has not actually helped so much anyways.

Instead, seize the day and identify clear classes of drivers. Take the guess work away from developers and settle on some solid drivers to handle specific issues. There should be a guide listing the types of things that are resolved by drivers and an active interest in encouraging the development of the different classes of drivers. 

Without this business minded approach there is only misunderstanding, missed opportunity, fragmentation and outdated dogma. One thing is sure, reactive data and state solutions are coming out of the woodwork and they can not be ignored and be forced into a second rate position in the Cycle framework. Everything finds it's natural plane of existence.

**Comment by ivan-kleshnin** [comment]
I've got an idea. We can leave drivers alone. What if we accept @staltz opinion and agree 
they have quite established place in Cycle ecosystem.

What we can do instead is generalize looping technique. 
Something like

```js
function main() {
  // ...
 
  Cycle.loop({ // need to support trees, not only plain records
    state: stateObserver,
    // and whatever we want to loop 
  })
}

Cycle.run(main, drivers) // calls Cycle.loop inside. 
// Can be force-checked to be plain record (sanity policy)
```

Maybe we are too much obsessed with `Cycle.run` and **single imperative call** statement.
Who said there can't be two or three? :wink:

**Comment by milankinen** [comment]
I agree. If the signal looping is not acceptable via drivers, then Cycle should at least provide a safe and convenient way to do it in the userland (like [TSERS does](https://github.com/tsers-js/core#run--input-input--output-loop--output)).

**Comment by staltz** [comment]
What is "unsafe" and "inconvenient" about a proxy Subject?

```js
let update = {
  users: new Subject(),
  articles: new Subject(),
  ...
}
```

You could on the other hand just use one Subject here...

**Comment by HighOnDrive** [comment]
@Staltz Never mind "unsafe" and "inconvenient" it's retarded. Cyclist want to work with real data not cheap low end versions of stores. Come on already, get with the times here! Lets start considering the bigger picture which is comprised of serious data backends for today's and tomorrows awesome apps. 

There must be good input and output for such data including streaming and also robust in client access, transmission and integration. Also the growth of Cycle will stay at a snails pace until data and state flow freely within Cycle. It also effects the construction of components and wiring them up in every way, how is that for a side-effect of blocking progress!

Maybe your just not aware of how your tunnel vision does not line up with the very real importance of data? At one point you said you were not a backend guy. Anyways, on every point where it concerns integrating Cycle apps together with data your thinking far to small for the real enterprise world. Will Cycle always be a toy?

**Comment by ivan-kleshnin** [comment]
> EDIT: @ivan-kleshnin my criticism about Cycle's read/write effect vagueness relates to the actual implementations of the official Cycle drivers, not the idea itself.

Ah. Got it. 

> What is "unsafe" and "inconvenient" about a proxy Subject?

It was late and I've messed up my code snippet. Updated now.

---

From the other hand: separate run and looping are going to repeat each other completely.

In **Haskell** mutable variables are implemented through `IO` becase common variables are absolutely immutable. So state is IO effect. Which goes absolutely inline with my wods and CycleJS ideas. Cc @zudov (I know you're aware of all Haskell, PureScript and CycleJS)

**Comment by milankinen** [comment]
Well said @ivan-kleshnin!

>What is "unsafe" and "inconvenient" about a proxy Subject?

For example memory management. If you want the architecture to be "fractal", you can't just create some Subjects and let them be like that. For "root" component it's ok, but how about if some other component want to use the component e.g. inside a list. Even **single** component like this can ruin your entire application. 

Using `using` you might say? If you want to use subjects with `using` you'll end up decomposing and switching streams. Good like teaching that kind of things to beginners.
```javascript
function main({DOM, HTTP, A, B, ...}) {
  const sinks$ = O.using(() => new Subject(), s => {
    const sinks = ...do somehing with subject...
    return sinks
  })
  return {
    DOM: sinks$.map(s => s.DOM).switch(),
    HTTP: sinks$.map(s => s.HTTP).switch(),
    A: sinks$.map(s => s.A).switch(),
    B: sinks$.map(s => s.B).switch(),
    ...
  }
}
```
And why my clicks are not sending any HTTP requests?! Oh. I forgot to add `.shareReplay(1)` to my `sinks$`...

**Comment by staltz** [comment]
@ivan-kleshnin 
> State changing is an effect nevertheless.

I don't think in the same way. Read my comment in this thread: https://github.com/cyclejs/core/issues/278#issuecomment-199810972

>You need to define what is the "external world" and what is the "application". It's pretty clear that graphics and mouse inputs are from the external world, but so is a server (in relation to a client app). Other not-so-obvious distinctions are e.g. storage. It's not actually external to the computer, but it is external to the application, so we can consider it part of the external world.
>On the other hand, we don't in Cycle.js consider e.g. the memory heap as part of the external world. That's why state in memory is not a side effect in Cycle apps.

Also in Elm applications state is stored in the `foldp` operator, which is equivalent to RxJS `scan`, and Elm is a functional language.

> Why it's an effect? Because there may be a write to localStorage.

Well state changing in memory is one thing, localStorage is something entirely else. The latter is really an effect.

> http://research.microsoft.com/en-us/um/people/simonpj/papers/marktoberdorf/mark.pdf

Good paper, thanks. I was already aware of Haskell's history and Stream I/O, there is a mention of it on the Cycle docs.

> Drivers can be renamed to Effects.

Not exactly. Drivers are adapters in order to interact with effects. Usually in Haskell and PureScript, *Effects are encoded as pure data* (monadic I/O), and that's why they call those things "Eff". But it's very different in Cycle. Drivers are **impure** functions. A driver is to an effect what a restaurant is to food. But food != restaurant.

@milankinen 
> For example memory management. If you want the architecture to be "fractal", you can't just create some Subjects and let them be like that.

With Subjects you can use `.takeUntil` or `using`, and memory management is solved.

> and of course using subjects in userland is so inelegant

Subjective. In other words, a state driver is IMO much more inelegant than a harmless Subject.

**Comment by HighOnDrive** [comment]
> That's why state in memory is not a side effect in Cycle apps.
VERSUS
>  It's about unnecessary complexity that could easily be avoided.

Boiled down to it's essence I just do not want to do state management with proxy Subjects and half baked options. I have a lot of data and want to work with it in a practical and elegant way. I do not want to jump through hurdles and distractions or have to contort and limit my design to less powerful paradigms like components. Indeed, this is what is great about CALMM-JS, they found a way to send atoms into and out of components. Cycle holds no candle to this achievement.

What is also interesting here is that Cycle says that it is a dataflow solution. Really? Are we not stepping on your own feet when this is said? Is this a half truth or the complete truth? I think everyone agrees that dataflow, in the way that Cycle uses the term, refers to sources and sinks cycling their way through the components of an app.

How ironic that now this DATAflow is being reserved only for side-effects. Something is not on the mark here, even deceptive, false advertizing for sure IMHO. If Cycle is about dataflow then why is dataflow being blocked by out moded arguments based on how everyone else has done it? I thought I also read that " it's very different in Cycle" in the previous post.

I think Cycle's major achievement to date is the simple architecture that involves a main() and circuits for processing effects, that is pretty wonderous. Now, because it is wonderous there are things that can be done with harnessing the powers of circuits, good ideas do come to mind. However, if the steward of the circuits is doing an inadequate job then the circuit machinery should be handed back to the people.

What really disappoints me is that Cycle is short sighted, limited by fundementalist dogma and just not thinking ahead and being where the market and therefore the developers actually are. I'm sure things will make way for greater ideas but it still sucks to always have to argue over such straightforward and helpful innovations.

"In the end, complacency is a symptom of mediocrity and mediocrity is the result of a leadership organization that chooses not to lead, but instead, to manage how to be better or more efficient around "what is" and not "what should be" or "what's next"." — INNOVATE OR DIE, What's the Future of Business?, Brian Solis

**Comment by staltz** [comment]
If you don't like it, use something else. You seem to have found it anyway :P

**Comment by ivan-kleshnin** [comment]
@staltz coupling and cohesion are always in complex contradictions to each other. At some point raising cohesion lead to a growing coupling. Or reducing coupling lead to a reduced cohesion.

It seems impossible to slice the state management on so thin "slices" that 

1. Every effect operation will be precisely in a driver 
2. Every logic will be precisely in the app
3. Number of data channels will be somewhat affordable.

We need to sacrifice something. You basically propose to sacrifice 3. but what I want is fair comparison.

IMO, despite that some ops don't literally "change the world" it's better to keep them with those ops which do. Every driver has some sort of logic. This imaginary "state driver" is not an exception. It's just a more vivid case.

There is a similar question of "logic-less templates vs logic-full templatess". People were trying to separate "logic" and "presentation" forever. Always ending with either presentation leaking to Controller or with logic leaking to View. Now we are pretty confident they can't be separated by this principle.

We can't distill "logic" or "effect" from anything completely IMO.
The split of interest is "business logic" vs "everything else". 

This is not to change your opinion. This is to explain my position and I probably won't continiue this discussion. No point. Time will judge who was right and wrong.

**Comment by HighOnDrive** [comment]
> If you don't like it, use something else. You seem to have found it anyway :P

That's just so unreasonable and just reflective of the box that Cycle has become confined to. When you learn how something does not work, then moving on to better more open things is good advice :100:

**Comment by staltz** [comment]
@ivan-kleshnin as a framework, it needs a set of guidelines (an opinion really) on what to do. In this sense, Cycle is a framework, not a library. You can use it as a library if you wish. If you know what you're doing, go ahead and make a state driver. The reason why I've been against that is:

> it's about conflating concepts. This starts making it less clear what is a driver for. The more drivers are used for ad hoc solutions, the more confusion appears as to "when should I make a driver and when should I not?". Exceptions like the noop are shortcuts which make it harder for beginners to understand what do drivers do. "well, you can safely use them as channels to transfer messages for whatever you want" is something I'd be uncomfortable teaching to other programmers.

**Comment by staltz** [comment]
As a reminder for myself and others, the reason why this issue is not a discussion is just because of this suggestion given a long time ago https://github.com/cyclejs/core/issues/204#issuecomment-189545542

**Comment by ivan-kleshnin** [comment]
@HighOnDrive every community acquires dogmas gradually. This is of human nature. 

@staltz at least we decided that CycleJS is a framework. 
I remember it was a questionable subject :wink:

**Comment by staltz** [comment]
But let's not "seal" this discussion. One thing that was left unaddressed is how to achieve a lens API that does rely on ad hoc drivers. That would be nice indeed. Also, it's not like state management is ignored in Cycle, the recommended approach (not documented, but I want to write a chapter on "large apps") is to map action object stream (from intent) to reducer functions each, and then accumulate them with scan. It's equivalent to the ideas in Elm architecture (update functions) and Redux (reducer functions).

If anyone has the energy for this, please think about a lens API without ad hoc drivers.

**Comment by ivan-kleshnin** [comment]
@staltz lenses won't help much without an added abstraction over state. Not saying it have to be a driver astraction but there should be something to absorb complexity. Otherwise the only thing we can do is to add some syntactic sugar.

### Granular state vs Monolithic state

https://github.com/ivan-kleshnin/cyclejs-examples/tree/master/x.a-form
vs
https://github.com/ivan-kleshnin/cyclejs-examples/tree/master/x.b-form

Conclusion: parity

### Lenses vs Manual
https://github.com/ivan-kleshnin/cyclejs-examples/tree/master/x.b-form
vs
https://github.com/ivan-kleshnin/cyclejs-examples/tree/master/x.c-form

Conclusion: lenses are neat

**Comment by whitecolor** [comment]
> at least we decided that CycleJS is a framework. 
I remember it was a questionable subject 

Cycle is functional javascript framework. To me it is much more framework then different angulars and other bloated stuff. And it is natural and quite comprehensive because allows to deal with everthing in unified cyclic/functional style.

@HighOnDrive Your arguments seem to me a little vague, it is not clear what you lack in cycle. You tried to implement stuff with cycle and met particular difficuties that hold you back? It seem that you had not really tried to make something with cycle. Instead of looking for better place for philosophical discussions, I would better try it out.

**Comment by HighOnDrive** [comment]
@whitecolor IMO it's dangerous to implement to much in Cycle yet, especially if what your working toward is important and has a shelf life beyond hacking. Just take a look back over the pass six months, we have isolate issues, VDOM troubles and wholesale replacement and kids going gaga over comparing the speed of computations and distracting Cycle from mature business endeavors, plus a leader who has not time to even write his article on larger apps. So you see, we're all trying Cycle out but if your honest and take the blinders off it's easy to see Cycle has become stuck in it's own lack of judgement.

So to reiterate I have a data centric app and I need to move that data around the app in many ways and access it with lenes. I shun making my components overly complex and instead aim to drive the app through data and dataflow (unlike React where you only think in terms of components). To do this properly it would be best to have a state-driver as already specified and implemented in posts above. At the end of the day I want what is unique about what Cycle brings to the table (dataflow) to be even better. The dial for greatness has room to be cranked some more, that all we're talking about here, reaching the potential.

**Comment by TylorS** [comment]
I don't think there is any issues blocking anyone from writing and using a model driver, I just don't believe it is necessary or fits with the general ideas that Cycle was built on. As a matter of fact I wrote a small one previously to toy around with; maybe its useful to someone here.

```js
function makeStateDriver(initialState = {}) {
  return function stateDriver(sink$) {
    const source$ = sink$
      .startWith(initialState)
      .scan(Object.assign)
      .replay(null, 1)
    source$.connect()
    return source$
  }
}

export {makeStateDriver}
```

If someone needs it to be a core part of the framework, than fork it and build it in Cycle-core is extremely small, it took me only a few hours to rebuild it in Python before I could even really read Javascript.. It took me less than an hour to rebuild it for motorcycle. Its easy to grok and should be easily extended with just about anything, not just limited to something that handles a model. I urge you all to do this for your own needs (and others), there is nothing wrong with that, but at that point it just isn't Cycle.js. 

I'm working on a Cycle.js project, which has a shelf life of, well forever, it's for a company whos only product *is* this application. The project with become **much** larger. Its at 6000 LOC now (not including the libraries we've extracted away), and no one on the team has had a single issue with the way we are doing things now, and are constantly saying "OMG this is so awesome". And to this point we haven't had to use a single proxy subject, though I do understand where and when we might need them. I'll also note that one of the team members, had never touched Cycle or Rx prior to this, and is probably the most productive person we have.

@HighOnDrive I'm sorry you feel that Cycle does not do, or see the same way as you would like. I hope you find somewhere that does.

**Comment by whitecolor** [comment]
@HighOnDrive 

I think I could(n't) be more sarcastic.

> IMO it's dangerous to implement to much in Cycle yet

Audaces fortuna juvat

> plus a leader who does not take the time to write his long promised article on larger apps.

Omnia tempus habent.

> So you see, we're all trying Cycle out but if your honest and take the blinders off it's easy to see that Cycle has become stuck in it's own lack of judgement.

Sic itur ad astra.

**Comment by HighOnDrive** [comment]
@TylorS You are echoing the underlying issue in saying that Cycle extremely small and the core idea can be replicated by anyone. At that point anyone can just pick a VDOM and a way to handle backend IO and be off to the races. 

So, does this mixing and matching produce a framework? Why is a framework even required then? Looks like Cycle is just an idea, a way of using RxJS and interfacing it to the DOM. FRP and RxJS what its all actually about then. There is also dataflow but it's potential seems to have been diminished by dogmatic choice.

Thanks also for sharing the state-driver snippet, I remember it from when we use to work on ideas to enhance Cycle, good that you still remember and funny how we are still dealing with the same haunts. That being so, making a state-driver is not the problem here, instead making Cycle a robust environment for development and using circuits for useful tasks is what this discussion is about.

As far as LOC goes my app has 6000 and extracted libraries beat, however, its not about lines of code, rather it should be about engineering and finding ways to reduce boilerplate and build apps in an efficient manner. As far as being productive, the developer you mention changes their approach every week (see gitter for the track record), seeing the light anew each day it seems. This is bound to happen until state management and a proper component strategy arrive (hopefully, right now most Cyclist are new or comfortably numb). Until then the motto seems to be, feel free to hack but never exceed the dogma.

>I'm sorry you feel that Cycle does not do, or see the same way as you would like.

Again, that's not what this discussion is about, the problem is that Cycle is not ready to build large apps. The focus is on regurgitating and reworking the same drivers and not putting in place anything to take you beyond building much more than another BMI example app.

> I hope you find somewhere that does.

Looks like your right about this one point. It looks like there are intelligent movers and shakers who have crafted solutions for the things that Cycle neglects. Like you said, making a little cycletron is the easy part. Then it's what you do with it that really matters and the improved options are looking fabulous!

@whitecolor Thanks for the forward looking Latin phrases, I'll keep them in my back pocket :rocket:

**Comment by whitecolor** [comment]
@HighOnDrive 
> Again, that's not what this discussion is about, the problem is that Cycle is not ready to build large apps. The focus is on regurgitating and reworking the same drivers and not putting in place anything to take you beyond building much more than another BMI example app.

OP was about particular dev problem, there are some dev ideas in the discussion in the thread, some sensible, though no solid yet, and I'm sure everthing will be considered. But I see that you (maybe not intentionally) tend to turn the dev discussion into some of vague form mutual altercations without profitable result.
Cycle is not going to build large apps for you, if you are not ready to build (large) apps with cycle - so much the worse for you.

**Comment by HighOnDrive** [comment]
@whitecolor To the contrary, I know how to build large apps and I'm just hoping Cycle can keep being the building blocks I use to build with. To me the profitable result is highlighting circuits and their importance and potential and getting my doubts out there, where they can be considered and remendied by a greater force of friends :smile:

**Comment by laszlokorte** [comment]
@HighOnDrive I really do not get what is your problem.

> So, does this mixing and matching produce a framework? Why is a framework even required then? Looks like Cycle is just an idea, a way of using RxJS and interfacing it to the DOM.

Does anyone force you to use cycle? Did anyone tell you that cycle was required framework for you to build apps?

> At that point anyone can just pick a VDOM and a way to handle backend IO and be off to the races.

Yes you can do exactly that and nobody will stop you. You can even use cycle as it is and build as many state drivers and lensing mechanisms as you like. nobody cares.

> plus a leader who does not take the time to write his long promised article on larger apps.

I have no idea why you are looking at @staltz as a leader. Actually I am not quite sure why you need a leader at all. We are a free community. Everyone has her and his own ideas. Staltz had one idea that he implemented, documented and promoted and some people seem to like it. For sure there are many other ideas out there. Some equally good, some not so good and some even better ones - depending on the use case.

If you have understood the concept of cycle (or even if you only think you have - it does not matter) you will lose nothing by not using cyclejs. Just go and implement your super large business application as you like. Those 100 lines of cyclejs code will not affect your overall architecture the slightest. Just replace them as you like.

So why are you looking for a "leader" in the first place? If I am writing that this or that is not how it should be done in cycle or if Staltz is arguing that drivers are not a place for doing some stuff it's just my opinion based on my understanding or his personal opinion based on some reasoning - nothing more than that. In Staltz' case that reasoning lead to the creation of cycle so if you like cycle I would argue that there is some value in trying to understand his reasoning but if you disagree with parts of it that's totally fine. You do not have to obey the great master of streams who leads mankind into a world without state drivers.

Who is stopping you from building you application with CycleJS+statedriver+lenses or with TSERS or with calmm? If those tools work that great I see no problem at all.

**Comment by HighOnDrive** [comment]
@laszlokorte Thanks and fine, the I am the leader reference is just not that hard to miss in the wild and from how typical exchanges in the Cycle world usually go. Not looking for no leader here, just dislike anyone standing in my way and cowards. Had a case of obligatory cobwebs there for a while, was feeling stuck in the muck.

**Comment by ivan-kleshnin** [comment]
@laszlokorte despite I agree with most you've said the thing is nobody wants to work in vacuum. 

If @HighOnDrive or other person goes against mainstream he kinda loses the purpose of using CycleJS. LOC quantity does not matter if your code is not 100% compatible upfront. Let's be honest, most people here use or going to use CycleJS for fun. When you select a project for a business your criterias change substantially. I'm also interested where CycleJS is going and how the top contributors envision the future.

We do not weigh a framework on it's own. I would say community is even more important.

**Comment by whitecolor** [comment]
> Let's be honest, most people here use or going to use CycleJS for fun. When you select a project for a business your criterias change substantially.

It took time before angular and react started to be used in production and become considered "business safe". And there is a little rub with cycle currently: its a hackers way, and most devs are not really hackers. But I sure things going to change in future (only hackers will left in the industry, others will be not needed, ha-ha).

@HighOnDrive to understand if cycle (or anything eles) feat you needs just open freaking editor and start coding, there is no other way. If you don't take any risks you are gaaranteed to fail.

**Comment by arlair** [comment]
From this issue:
> (I don't work on Cycle.js full time), but I'm actively search for time to work on issues. Of course it'd be nice to have more contributors from the community.

Whilst I don't necessarily agree with (or even understand ;) ), everything in Cycle.js, I for one certainly appreciate what @staltz and others in open source provide for us and have helped me learn so far. Perhaps there are some misconceptions about expectations and entitlements for freely available open source projects and other peoples time and what they do with it. A common theme in this age of entitlement.

Having followed Cycle.js from the start, I sometimes think about how much Cycle.js _can_ do when compared to other frameworks with much more resources.

A technique I have experimented with is attempting to write the top level code in the simplest and most direct way for that particular task, trying to put existing ideas in the back of my mind. This can lead to fresh new ideas that can be worked back into the existing code.

Sometimes people complain things are changing too often, other times things are stuck in the same place (sometimes this is the same person). I'm not sure of the answer to this particular problem, hopefully an acceptable consensus will emerge soon, but it is nice to see people contributing their ideas :)

**Comment by ronag** [comment]
Not that this isn't an interesting and important discussion or anything. But I think all of this is irrelevant for the original issue. Which seems to have been forgotten.

May I suggest we close this issue. Open a new one for the original issue. And open another one for this discussion?

**Comment by Cmdv** [comment]
Here you go guys https://github.com/cyclejs/core/issues/280

**Comment by HighOnDrive** [comment]
@whitecolor This is not really about state handling so I'll reply here :-)

> to understand if cycle (or anything else) feats you needs you need to just open freaking editor and start coding, there is no other way. If you don't take any risks you are gaaranteed to fail.

There is actually so much you can gain without coding and finding ways to glean a lot of knowledge before getting to a editor is good too, of course this works best if you have a lot of direct hands on coding under your belt already. 

Being so I first like to architect and engineer solutions, often in diverse areas including graphics, audio, mathematical, backend, etc. I can already tell what works and does not just by having a good look around. I am primarily concerned in fusing together products and do not have so much time for playing around.

My primary task as I see it is to manage risk once I have the vision for something worthwhile in hand. I take great risks in my life for the passions I have, by this point in life I have a purpose to my comings and goings. Vigilance, that's the usual word of the day for me as live the high risk life most others would run far away from.

You need experience to know better and to know better you need experience. So how you do things all really depends on where your at along this continuum of experience. I whole-heartedly agree that taking risks is the way to go (cowards are the first ones listed as going to hell) but you need to pick your battles too, so you can win the war. Now I got to go write some code, cheers!

**Comment by Cmdv** [comment]
@HighOnDrive not really about the actual issue either, I tried to name it best I could to just shift the whole conversation over there. Because even though this conversation has been deep etc it has nothing to do with the original issue. If people could continue over on that discussion that would be appreciated :)

Probably too late I know but I'm interested in the original issue and trying to solve that :+1:
